source/validate_cfg.cpp

```diff
@@ -244,30 +254,154 @@ spv_result_t FirstBlockAssert(ValidationState_t& _, uint32_t target) {
+      break;
+    default:
+      assert(1 == 0 && "Not defined type");
+  }
+  // TODO(umar): Add header block for continue constructs to error message
+  return "The " + construct_name + " construct with the " + header_name + " " +
+         header_string + " does not " + dominate_text + " the " + exit_name +
+         " " + exit_string;
+}
+
 spv_result_t PerformCfgChecks(ValidationState_t& _) {
   for (auto& function : _.get_functions()) {
+    // Check all referenced blocks are defined within a function
+    if (function.get_undefined_block_count() != 0) {
+      string undef_blocks("{");
+      for (auto undefined_block : function.get_undefined_blocks()) {
+        undef_blocks += _.getIdName(undefined_block) + " ";
+      }
+      return _.diag(SPV_ERROR_INVALID_CFG)
+             << "Block(s) " << undef_blocks << "\b}"
+             << " are referenced but not defined in function "
+             << _.getIdName(function.get_id());
+    }
+
     // Updates each blocks immediate dominators
     vector<const BasicBlock*> postorder;
+    vector<const BasicBlock*> postdom_postorder;
     vector<pair<uint32_t, uint32_t>> back_edges;
     if (auto* first_block = function.get_first_block()) {
+      /// calculate dominators
       DepthFirstTraversal(*first_block, successor, [](cbb_ptr) {},
                           [&](cbb_ptr b) { postorder.push_back(b); },
                           [&](cbb_ptr from, cbb_ptr to) {
                             back_edges.emplace_back(from->get_id(),
                                                     to->get_id());
                           });
-      auto edges = libspirv::CalculateDominators(postorder);
-      libspirv::UpdateImmediateDominators(edges);
+      auto edges = libspirv::CalculateDominators(postorder, predecessor);
+      libspirv::UpdateImmediateDominators(
+          edges, [](bb_ptr block, bb_ptr dominator) {
+            block->SetImmediateDominator(dominator);
+          });
+
+      /// calculate post dominators
+      auto exit_block = function.get_pseudo_exit_block();
+      DepthFirstTraversal(*exit_block, predecessor, [](cbb_ptr) {},
+                          [&](cbb_ptr b) { postdom_postorder.push_back(b); },
+                          [&](cbb_ptr, cbb_ptr) {});
+      auto postdom_edges =
+          libspirv::CalculateDominators(postdom_postorder, successor);
+      libspirv::UpdateImmediateDominators(
+          postdom_edges, [](bb_ptr block, bb_ptr dominator) {
+            block->SetImmediatePostDominator(dominator);
+          });
     }
 
+    /// Check all backedges target only loop headers and have exactly one
+    /// back-edge branching to it
+    set<uint32_t> loop_headers;
+    for (auto back_edge : back_edges) {
+      uint32_t back_edge_block;
+      uint32_t header_block;
+      tie(back_edge_block, header_block) = back_edge;
+      if (!function.IsBlockType(header_block, kBlockTypeLoop)) {
```

Comment:
Structured control flow is a requirement only if the Shader capability is declared.
Subtle point. :-)